src/key_store/mod.rs

```diff
@@ -0,0 +1,223 @@
+    /// corresponding `KeyPackage`. This removes the `KeyPackageBundle` from the
+    /// store. Returns an error if no `KeyPackageBundle` can be found
+    /// corresponding to the given `KeyPackage` hash.
+    pub fn take_key_package_bundle(&self, kp_hash: &[u8]) -> Option<KeyPackageBundle> {
+        // We unwrap here, because the two functions claiming a write lock on
+        // `init_key_package_bundles` (this one and `generate_key_package`) only
+        // hold the lock very briefly and should not panic during that period.
+        let mut kpbs = self.init_key_package_bundles.write().unwrap();
+        kpbs.remove(kp_hash)
+    }
+
+    /// Retrieve a `CbGuard` from the key store given the `SignaturePublicKey`
+    /// of the corresponding `Credential`. The `CbGuard` can be dereferenced to
+    /// obtain an `RwLock` on the desired `CredentialBundle`. Returns an error
+    /// if no `CredentialBundle` can be found corresponding to the given
+    /// `SignaturePublicKey`.
+    pub(crate) fn get_credential_bundle<'key_store>(
+        &'key_store self,
+        signature_public_key: &'key_store SignaturePublicKey,
+    ) -> Option<CbGuard> {
+        let cbs = self.credential_bundles.read().unwrap();
+        if !cbs.contains_key(signature_public_key) {
+            return None;
+        }
+        Some(CbGuard {
+            cbs,
+            index: signature_public_key,
+        })
+    }
+
+    /// Generate a fresh `KeyPackageBundle` with the given parameters, store it
+    /// in the `KeyStore` and return the corresponding `KeyPackage`. Throws an
+    /// error if no `CredentialBundle` can be found in the `KeyStore`
+    /// corresponding to the given `Credential` or if an error occurred during
+    /// the creation of the `KeyPackageBundle`.
+    pub fn generate_key_package(
+        &self,
+        ciphersuites: &[CiphersuiteName],
+        credential: &Credential,
+        extensions: Vec<Box<dyn Extension>>,
+    ) -> Result<KeyPackage, KeyStoreError> {
+        let credential_bundle = self
+            .get_credential_bundle(credential.signature_key())
+            .ok_or(KeyStoreError::NoMatchingCredentialBundle)?;
+        let kpb = KeyPackageBundle::new(ciphersuites, &credential_bundle, extensions)?;
+        let kp = kpb.key_package().clone();
+        // We unwrap here, because the two functions claiming write locks on
+        // `init_key_package_bundles` (this one and `take_key_package_bundle`)
+        // only hold the lock very briefly and should not panic during that
+        // period.
+        let mut kpbs = self.init_key_package_bundles.write().unwrap();
+        kpbs.insert(kp.hash(), kpb);
+        Ok(kp)
+    }
+
+    /// Generate a fresh `CredentialBundle` with the given parameters and store
+    /// it in the `KeyStore`. Returns the corresponding `Credential` or an error
+    /// if the creation of the `CredentialBundle` fails.
+    pub fn generate_credential(
```

Comment:
It is called `generate_credential` but returns a `CredentialBundle` that's confusing.

Reply:
Also, why only return the credential?

Reply:
That should be a generates, not returns in the first comment.
My issue is not with returning a credential, but you don't generate a credential but a credential bundle. That's unexpected.

Reply:
I see. So `generate_credential_bundle` (and `_key_package_bundle`) as a replacement? The documentation should then explain why it only returns a `Credential` (or `KeyPackage` respectively).

Reply:
Both isn't really great. I don't have a good suggestion for how to resolve this. But `generate_credential_bundle` is at least doing what it says. You could also have a separate generate and get function, i.e. don't return anything. But really this is some sort of `generate_key_pair`, not a `generate_random_point`, so this has to be made clear.
But since the API will change anyway I'm also fine with an explicit follow up issue.